feat(enhancement): Add "timer" trigger for missions (#9089)
endless-sky/endless-sky: tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt

# Copyright (c) 2023 by Hurleveur
#
# Endless Sky is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later version.
#
# Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program. If not, see <https://www.gnu.org/licenses/>.

test-data "Timer Test Mission"
	category mission
	contents

> 		mission "10 second timer"

Reviewer: ```suggestion
		mission "1 second timer"
```

			landing
			timer "Why do I have a name?"

> 				time 600

Reviewer: ```suggestion
				time 60
```

				reset "leave system"
				on timeup
					set "test: timer condition"

> test "Timer - Simple 10 seconds timer"

Reviewer: ```suggestion
test "Timer - Simple 1 second timer"
```

	status active

> 	description "Tests if a timer will trigger a mission event in 10 seconds as planned."

Reviewer: ```suggestion
	description "Tests if a timer will trigger a mission event in one second as planned."
```

	sequence
		inject "Three Earthly Barges Save"
		inject "Timer Test Mission"
		call "Load First Savegame"
		# Trigger the mission
		call "Depart"
		call "Land"
		call "Depart"
		apply

> 			"test: steps to wait" = 30 * 60

Reviewer: ```suggestion
			"test: steps to wait" = 20 * 60
```

> 			"test: steps to wait" = 20 * 60

Reviewer: ```suggestion
			"test: steps to wait" = 2 * 60
```

		label "floating"

> 		# An empty input to make 30 seconds pass.

Reviewer: ```suggestion
		# An empty input to make 20 seconds pass.
```

> 		# An empty input to make 20 seconds pass.

Reviewer: ```suggestion
		# An empty input to make two seconds pass.
```

> 		# empty input to make 10 seconds and one frame pass

Reviewer 2: this comment seems wrong?
also maybe give it a bit more time, just to be sure?

Author: Any idea how much? I'm afraid I've got no experience with the tests (these were actually written by someone else—I forget who, offhand).

Author: Boosted it to 30s.

# Copyright (c) 2023 by Hurleveur
#
# Endless Sky is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later version.
#
# Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program. If not, see <https://www.gnu.org/licenses/>.

test-data "Timer Test Mission"
	category mission
	contents
		mission "10 second timer"
			landing
			timer "why do I have a name"
				time 600
				reset "leave system"
				on timeup
					set "test: timer condition"

test "Timer - Simple 10 seconds timer"
	status active
	description "Tests if a timer will trigger a mission event in 10 seconds as planned."
	sequence
		inject "Three Earthly Barges Save"
		inject "Timer Test Mission"
		call "Load First Savegame"
		# Trigger the mission
		call "Depart"
		call "Land"
		call "Depart"

> 		watchdog 10000

Reviewer 2: not supposed to be here

Author: Just the `watchdog` line?

Reviewer 2: yes

test-data "Timer Test Mission"
	category mission
	contents
		mission "short timer"
			landing
			repeat
			timer 60
				on timeup
					set "test: 1 second timer"
		mission "long timer"
			landing
			repeat
			timer 36000
				on timeup
					set "test: 10 minute timer"
		mission "optional long timer"
			landing
			repeat
			timer 36000
				optional
				on timeup
					set "test: optional 10 minute timer"

test "Timer - Simple 1 second timer"
	status active
	description "Tests if a timer will trigger a mission event in one second as planned."
	sequence
		inject "Three Earthly Barges Save"
		inject "Timer Test Mission"
		call "Load First Savegame"
		# Trigger the mission
		call "Depart"
		call "Land"
		call "Depart"
		apply
			"test: steps to wait" = 2 * 60
		label "floating"
		# An empty input to make two seconds pass.
		input
		apply
			"test: steps to wait" -= 1
		branch "floating"
			has "test: steps to wait"
		assert
			has "test: 1 second timer"
			not "test: 10 minute timer"
			not "test: optional 10 minute timer"
		call "Land"
		# This works when I test it manually.
#		assert
#			has "short timer: done"
#			not "long timer: done"

> #			has "optional long timer: done"

Reviewer 3: Yeah, looks like that was it.
